### ISSUE TITLE
Many unknowns classification and poor training results

### ISSUE THREAD

**Issue by MahdiElHadj** [AI4Math-ShanZhang/OWOBJ#4]
Dear author,
Thank you for sharing your source code.
Over the past few days, I have been testing OWOBJ on my custom dataset, but I am encountering a large number of unknown predictions. The model also seems unable to correctly classify samples within the unknown category.
After 15 epochs, I am obtaining the following results, which are quite low:
```
Current class AP50: tensor(2.)
Current class Precision50: 4.036036036036036
Current class Recall50: 0.05708230790265245

Known AP50: tensor(2.)
Known Precision50: 4.036036036036036
Known Recall50: 0.05708230790265245

Unknown AP50: tensor(0.0010)
Unknown Precision50: 0.010065931853641352
Unknown Recall50: 0.03255738238645613
```

Do you have any suggestions regarding what the root cause might be, or which parameters I could adjust?
I have already tried balancing the weights between the different losses. I am considering introducing a decay parameter for the Objectness loss and the KL loss — do you think this could help?
Thank you in advance for your guidance.

**Comment by AI4Math-ShanZhang** [comment]
Hi MahdiElHadj,

Looking at your results, the known class AP50 is also quite low (2.0), which  may indicate the issue is not simply related to the objectness score modeling, and the base detection itself has not converged yet. 

Could you share what domain your custom dataset is from? If the data is confidential and cannot post in the public website, you are welcome to send 1–2 example images to my email so we can see and discuss. Understanding the domain and visual characteristics would help narrow down the cause.

For practical suggestions, 1. start by disabling the energy loss. Set --obj_engry 0.0 and verify that the base detector (known classes) trains properly on its own; 2. Tune the objectness temperature. The default --obj_temp 1.3 was calibrated for the M-OWOD benchmark. A wrong temperature for your dataset's feature scale can kill the gradient signal for the objectness head. Try values in the range 0.8–2.0 once the base detector is working; 3.  I would  suggest visualizing the detection results at different epochs to see how predictions evolve, and also check the cross-attention maps of queries and the feature map. When I meet the training issue, I like see the visualizions to debug. 

bests,
Shan

**Comment by MahdiElHadj** [comment]
Hello Shan,
Thank you for your quick reply.
At the moment, I’m unable to provide sample images from my dataset. However, I am currently exploring your suggested parameter adjustments for adapting the model to my custom dataset.
While reviewing the forward function in the PostProcess class, I noticed that some lines related to the temperature (temp) parameter are commented out. This seems to disable the effect of the temperature during evaluation/inference, while it is still applied during training.
Is this [behavior ](https://github.com/AI4Math-ShanZhang/OWOBJ/blob/f5c583e39593168e2313c149ba69801d79619f42/models/prob_deformable_sketch_detr.py#L649)intentional and part of the model’s design, or is it an oversight?
Here is the exact reference in the code: 
`        # obj_prob = torch.exp(-self.temperature*pred_obj).unsqueeze(-1)`
`        prob = out_logits.sigmoid()#*obj_prob`

Thank you again for your help!
Best,
Mahdi
